base/random.jl

```diff
@@ -253,14 +253,10 @@ rand(r::MersenneTwister, ::Type{Int128})  = reinterpret(Int128, rand(r, UInt128)
 rand{T<:Real}(r::AbstractRNG, ::Type{Complex{T}}) = complex(rand(r, T), rand(r, T))
 
 # random Char values
-# use simple rejection sampling over valid Char codepoint range
+# returns a random valid Unicode scalar value (i.e. 0 - 0xd7ff, 0xe000 - # 0x10ffff)
 function rand(r::AbstractRNG, ::Type{Char})
-    while true
-        c = rand(0x00000000:0x0010fffd)
-        if is_valid_char(c)
-            return reinterpret(Char,c)
-        end
-    end
+    c = rand(0x00000000:0x0010f7ff)
```

Comment:
I'm not sure I totally understand your question, but the change Jake made in 7efbc01 isn't really specific to ranges or the RNG. It's an idiom for functions that interact with global state (like the GLOBAL_RNG or STDOUT).  For example, the print function has one fallback method:

```jl
print(x) = print(STDOUT, x)
```

And then all custom types define methods `print(io::IO, x::MyType)`, and within those methods they should only call the two-arg print methods, explicitly using `io` instead of defaulting back to STDOUT.  This allows you to easily collect the output in, e.g., an IOBuffer without worrying about stuff leaking out to STDOUT.

The same applies to RNGs, since they are also a global that mutates.